### PR TITLE
catch nodesource runtime exceptions to be transmitted in the rm porta…

### PR DESCRIPTION
**Pull request by Aminelouati**
…l client

### CODE REVIEW

File: rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/RMRest.java [hunk 1]
```diff
@@ -420,8 +425,33 @@ public boolean createNodeSource(@HeaderParam("sessionid") String sessionId,
...
-        return rm.createNodeSource(nodeSourceName, infrastructureType, infraParams, policyType, policyParams)
-                 .getBooleanValue();
+        try {
+            nsState.setResult(rm.createNodeSource(nodeSourceName,
+                                                  infrastructureType,
+                                                  infraParams,
+                                                  policyType,
+                                                  policyParams)
+                                .getBooleanValue());
+            return nsState;
```
Comment (marcocast) [inline rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/RMRest.java]: dont return here and in the catch, just add a finally block after and return it there.

File: rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/RMRest.java [hunk 2]
```diff
@@ -420,8 +425,33 @@ public boolean createNodeSource(@HeaderParam("sessionid") String sessionId,
...
+            nsState.setErrorMessage(cleanDisplayedErrorMessage(ex.getMessage()));
+            nsState.setStackTrace(getStackTrace(ex));
+            return nsState;
+        }
+
+    }
+
+    private String cleanDisplayedErrorMessage(String errorMessage) {
+        String regex = "^[^:]*:(.*)";
+        Pattern pattern = Pattern.compile(regex);
```
Comment (marcocast) [inline rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/RMRest.java]: This pattern can be created only one time at constructor level

File: rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/RMRest.java [hunk 3]
```diff
@@ -420,8 +425,33 @@ public boolean createNodeSource(@HeaderParam("sessionid") String sessionId,
...
+        Matcher matcher = pattern.matcher(errorMessage);
+        if (errorMessage.matches(regex) && matcher.find()) {
```
Comment (marcocast) [inline rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/RMRest.java]: no need to errorMessage.matches(regex) as you already have matcher.find()

File: rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/RMRest.java [hunk 4]
```diff
@@ -420,8 +425,33 @@ public boolean createNodeSource(@HeaderParam("sessionid") String sessionId,
...
+            if (errorMessage.matches(regex)) {
```
Comment (marcocast) [inline rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/RMRest.java]: no need for this if

File: rm/rm-client/src/main/java/org/ow2/proactive/resourcemanager/common/NSState.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+/**
+ * @author ActiveEon Team
+ * @since 30/11/2017
+ */
+
+@PublicAPI
+@XmlRootElement
+public class NSState implements Serializable {
+
+    private boolean result;
```
Comment (marcocast) [inline rm/rm-client/src/main/java/org/ow2/proactive/resourcemanager/common/NSState.java]: better to have this object as an immutable object. 
To do that you need to :  
1 - make all its properties private final
2 - assign values to its properties in the constructor

To make it easier to create this object you can finally create a builder object with readable methods.

File: rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/RMRest.java (hunk 2 above)
Comment (fviale) [reply rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/RMRest.java]: pattern should be private static final

**Comment by fviale** [comment]
Retest this please

**Comment by fviale** [comment]
Currently, with this change, deploying a node source through the command line fails. It is thus necessary to fix this before merging

File: rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/json/NSStateView.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * If needed, contact us to obtain a release under GPL Version 2 or 3
+ * or a different license than the AGPL.
+ */
+package org.ow2.proactive_grid_cloud_portal.cli.json;
```
Comment (sonarqube-activeeon) [inline rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/json/NSStateView.java]: MINOR Rename this package name to match the regular expression '^[a-z]+(\.[a-z][a-z0-9]*)*$'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00120)

File: rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/RMRest.java [hunk 7]
```diff
@@ -420,8 +427,31 @@ public boolean createNodeSource(@HeaderParam("sessionid") String sessionId,
...
+                                                  policyType,
+                                                  policyParams)
+                                .getBooleanValue());
+        } catch (RuntimeException ex) {
+            nsState.setResult(false);
+            nsState.setErrorMessage(cleanDisplayedErrorMessage(ex.getMessage()));
+            nsState.setStackTrace(getStackTrace(ex));
+
+        } finally {
+            return nsState;
```
Comment (sonarqube-activeeon) [inline rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/RMRest.java]: MAJOR Remove this return statement from this finally block. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1143)

File: rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/RMRest.java [hunk 8]
```diff
@@ -115,6 +120,8 @@
 
     private SessionStore sessionStore = SharedSessionStore.getInstance();
 
+    private static final String REGEX = "^[^:]*:(.*)";;
```
Comment (sonarqube-activeeon) [inline rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/RMRest.java]: MINOR Remove this empty statement. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AEmptyStatementUsageCheck)

**Comment by sonarqube-activeeon** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by fviale** [comment]
Retest this please

File: rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/RMRest.java [hunk 9]
```diff
@@ -420,8 +427,31 @@ public boolean createNodeSource(@HeaderParam("sessionid") String sessionId,
...
+            nsState.setResult(rm.createNodeSource(nodeSourceName,
+                                                  infrastructureType,
+                                                  infraParams,
+                                                  policyType,
+                                                  policyParams)
+                                .getBooleanValue());
+        } catch (RuntimeException ex) {
+            nsState.setResult(false);
+            nsState.setErrorMessage(cleanDisplayedErrorMessage(ex.getMessage()));
+            nsState.setStackTrace(getStackTrace(ex));
```
Comment (fviale) [inline rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/RMRest.java]: beware of issues when converting a stack trace to json, the stack trace should be "escaped".

File: rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/RMRest.java [hunk 10]
```diff
@@ -420,8 +427,31 @@ public boolean createNodeSource(@HeaderParam("sessionid") String sessionId,
...
+
+        } finally {
+            return nsState;
+        }
+
+    }
+
+    private String cleanDisplayedErrorMessage(String errorMessage) {
+        Pattern pattern = Pattern.compile(REGEX);
```
Comment (marcocast) [inline rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/RMRest.java]: this should be final static in top of the class

File: rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/RMRest.java [hunk 11]
```diff
@@ -420,8 +428,30 @@ public boolean createNodeSource(@HeaderParam("sessionid") String sessionId,
...
+                                                  policyType,
+                                                  policyParams)
+                                .getBooleanValue());
+        } catch (RuntimeException ex) {
+            nsState.setResult(false);
+            nsState.setErrorMessage(cleanDisplayedErrorMessage(ex.getMessage()));
+            nsState.setStackTrace(StringEscapeUtils.escapeJson(getStackTrace(ex)));
+
+        } finally {
+            return nsState;
```
Comment (sonarqube-activeeon) [inline rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/RMRest.java]: MAJOR Remove this return statement from this finally block. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1143)

**Comment by sonarqube-activeeon** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**